plugin/trino-redshift/src/main/java/io/trino/plugin/redshift/RedshiftUnloadSplitSource.java

```diff
@@ -0,0 +1,115 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.plugin.redshift;
+
+import com.amazon.redshift.util.RedshiftException;
+import com.google.common.collect.ImmutableList;
+import io.airlift.log.Logger;
+import io.trino.spi.connector.ConnectorSplit;
+import io.trino.spi.connector.ConnectorSplitSource;
+
+import java.sql.Connection;
+import java.sql.PreparedStatement;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+
+import static java.util.Objects.requireNonNull;
+
+public class RedshiftUnloadSplitSource
+        implements ConnectorSplitSource
+{
+    private static final Logger log = Logger.get(RedshiftUnloadSplitSource.class);
+    private static final String RESULT_FILES_QUERY = "SELECT rtrim(path) FROM stl_unload_log WHERE query=pg_last_query_id() ORDER BY path";
+
+    private final Connection connection;
+    private final CompletableFuture<Boolean> resultSetFuture;
+    private final ConnectorSplitSource fallbackSplitSource;
+
+    private boolean finished;
+
+    public RedshiftUnloadSplitSource(ExecutorService executor, Connection connection, PreparedStatement statement, ConnectorSplitSource fallbackSplitSource)
+    {
+        requireNonNull(executor, "executor is null");
+        this.connection = requireNonNull(connection, "connection is null");
+        this.fallbackSplitSource = requireNonNull(fallbackSplitSource, "fallbackSplitSource is null");
+        requireNonNull(statement, "statement is null");
+        resultSetFuture = CompletableFuture.supplyAsync(() -> {
+            log.debug("Executing: %s", statement);
+            try {
+                connection.setReadOnly(false);
+                return statement.execute();
+            }
+            catch (SQLException e) {
+                throw new RuntimeException(e);
+            }
+        }, executor);
+    }
+
+    @Override
+    public CompletableFuture<ConnectorSplitBatch> getNextBatch(int maxSize)
+    {
+        return resultSetFuture
+                .thenApply(_ -> {
+                    ConnectorSplitBatch connectorSplitBatch = new ConnectorSplitBatch(getPaths().stream()
+                            .map(path -> (ConnectorSplit) new RedshiftUnloadSplit(path))
+                            .toList(), true);
+                    finished = true;
+                    return connectorSplitBatch;
+                })
+                .exceptionally(e -> {
+                    if (e.getCause() != null && e.getCause().getCause() != null && e.getCause().getCause() instanceof RedshiftException) {
+                        try {
+                            log.debug("Unload query execution failed. Falling back to using JDBC");
+                            return fallbackSplitSource.getNextBatch(maxSize).get();
```

Comment:
I don't like the idea of silently falling back here.
I would rather we fail the query and treat it as a bug. We should try to ensure that all unsupported cases are handled in `isUnloadSupported`
User can use session property or disable feature to get past this

Reply:
I see your point. I removed fallback and throwing in case of failure.